Basic disk control and M3U support

dszakallas opened the pull request:
This PR upgrades libretro header to a newer version that has extended disk tray interface and implements the disk tray control callbacks required for swapping disks.

Eject does nothing, insert saves the nvram, resets the emulation, loads the image and finally the nvram.

The purpose of this PR is to provide convenience functionality of swapping disks from inside retroarch, and m3u support, not faithful emulation of disk control of the hardware.

dszakallas committed:
naive disc tray emulation

dszakallas committed:
upgrade disk interface

dszakallas commented:
#14 #209

trapexit commented:
Will take a look. Thanks.

trapexit commented on libretro.c:
Not a fan of the change in style.

dszakallas replied:
do you mean the brace style, and the return value and static qualifier being on a separate line?

trapexit replied:
Yes

dszakallas replied:
Could you check again please?

trapexit replied:
Could you just squash all the commits you've made?

trapexit commented on libretro.c:
```diff
@@ -38,9 +38,73 @@
+static struct retro_disk_control_callback retro_disk_control_cb;
+static struct retro_disk_control_ext_callback retro_disk_control_ext_cb;
+
+static bool read_m3u(const char *file)
```
There isn't a generic m3u reader that could be used? Is this code picked up from another core?

dszakallas replied:
I couldn't find one.

trapexit commented on libretro.c:
```diff
@@ -38,9 +38,81 @@
+{
+  char line[PATH_MAX];
+  char name[PATH_MAX];
+  FILE *f = fopen(file, "r");
```
Should probably be using libretro functions if you can.

https://github.com/libretro/Genesis-Plus-GX/blob/b063aa81b90682ca89ae6fd1e958bc589b4ce34d/libretro/libretro.c#L3416

trapexit commented on libretro.c:
```diff
@@ -38,9 +38,81 @@
+// disk swapping
+#define M3U_MAX_FILE 4
+
+static char disk_paths[M3U_MAX_FILE][PATH_MAX_LENGTH];
```
The CD stuff needs to be moved over but the g_CAPS syntax is generally used for globals.

trapexit commented on libretro.c:
```diff
@@ -295,61 +367,264 @@ print_cdimage_open_fail(const char *path_)
   int rv;
 
-  game_info_path_save(info_);
+  extract_basename(g_GAME_NAME, info_->path, sizeof(g_GAME_NAME));
```
The name here is misleading? It's the path. Also... can't info be NULL? Opera allows starting without a game.

dszakallas committed:
Merge remote-tracking branch 'origin/master' into naive-disk-control

dszakallas committed:
naive disc tray emulation

Implement disc swapping support with M3U playlist parsing.

Changes address reviewer feedback from PR #211:
- Use lowercase variable names (cdimage, cdimage_sector, game_name)
- Use libretro utility functions for file I/O and string operations
- Add comprehensive null checks and bounds validation
- Support extended disk control interface when available
- Follow project coding style with K&R brace placement
